Introduce experimental FX library

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42741 Introduce experimental FX library**

**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
import torch
from torch.fx import GraphModule

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

Commit:
Introduce experimental FX library

[ghstack-poisoned]

Commit:
Update on "Introduce experimental FX library"


Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

Commit:
Update on "Introduce experimental FX library"


Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

Commit:
Update on "Introduce experimental FX library"


**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

Commit:
Update on "Introduce experimental FX library"


**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

vadimkantorov commented on torch/fx/graph_module.py:
```diff
@@ -0,0 +1,459 @@
+    'div': '{} / {}',
+    'mod': '{} % {}',
+    'pow': '{} ** {}',
+    'lshift': '{} < {}',
```
Should it not be << here? And below >>?

jamesr66a replied:
Yes, thanks

Commit:
Update on "Introduce experimental FX library"


**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

dzhulgakov commented on test/test_fx.py:
```diff
@@ -0,0 +1,61 @@
+
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+class TestFX(TestCase):
```
can you add test with passing dictionary explicitly as the argument, indexing it and constructing a dictionary in graph module (the most common cases)

jamesr66a replied:
Added a test that exercises what I think you're describing, please take a look

dzhulgakov commented on test/test_fx.py:
```diff
@@ -0,0 +1,61 @@
+
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+class TestFX(TestCase):
```
does scripting of a GraphModule work or not yet? (curious whether `inspect` works on generated forward)

zdevito replied:
It does work, see `eval_with_source` for how we inject the source information in a way such that inspect can find it.

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -0,0 +1,459 @@
+    def __iter__(self):
+        frame = inspect.currentframe()
+        calling_frame = frame.f_back
+        inst = list(dis.get_instructions(calling_frame.f_code))[calling_frame.f_lasti // 2]
```
yeah this code is super confusing, who wrote this!? (I wrote this...) Normally, unpacking a tuple would try to call `__iter__`. In most cases, we want to error on calling `__iter__` because the length of the symbolic value being iterated is unknown. However, in the pattern matching case `x, y = rhs` we know `rhs` must have length 2, otherwise this will error. This code snoops the callers code to see if an unpack is taking place and that there are two elements being matched. It then returns two things in the iterator to make sure the match succeeds. This makes things like `torch.maxi` which returns two values traceable when it otherwise wouldn't be. However, it isn't strictly needed since you can always rewrite the calling code to explicitly match `x = rhs[0]; y = rhs[1]`.

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -0,0 +1,459 @@
+    else:
+        return a
+
+class Graph:
```
We should move Graph to its own file (along with Node, etc.).

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -0,0 +1,459 @@
+    def __init__(self, root, graph=None, is_leaf_module=is_leaf_module):
+        super().__init__()
+        self.root = root
+        self.graph = graph if graph is not None else _run_symbolic_forward(root, is_leaf_module)
```
Rather than have `_run_symbolic_forward` be called in the ctor, we should have the api be `torch.fx.symbolic_trace(model) -> GraphModule`. This allows us to put the `_run_symbolic_forward` machinery in its own file, making it clear that `GraphModule` is separable from it. `GraphModule` can then just explictly take a Graph as argument. This makes it clearer how to use in transforms: GraphModule can be constructed given a Graph and an existing root module. `symbolic_trace` can generate a Graph from a module and then create a GraphModule from the parts.

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -0,0 +1,459 @@
+import torch
```
The PR comment is great documentation for this module, can you put it in the docstrings for `torch.fx` so we keep it around?

zdevito reviewed (approved):
Nice! I just have minor organization comments.

houseroad reviewed (commented):
Also a general question, will we use this file as the source of the truth? :-)

cc: @jamesr66a @zdevito @dzhulgakov

Commit:
Update on "Introduce experimental FX library"


**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
import torch
from torch.fx import GraphModule

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

Commit:
Update on "Introduce experimental FX library"


**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
import torch
from torch.fx import GraphModule

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

Commit:
Update on "Introduce experimental FX library"


**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
import torch
from torch.fx import GraphModule

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

jamesr66a commented:
> Also a general question, will we use this file as the source of the truth? :-)

Yes, that's the intent!

Commit:
Update on "Introduce experimental FX library"


**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
import torch
from torch.fx import GraphModule

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

rdzhabarov commented on torch/fx/symbolic_trace.py:
```diff
@@ -0,0 +1,84 @@
+    # so we can't call this function normally, otherwise it would try to unpack them
+    # instead, let's make python think that args and kwargs are normay variables
+
+def is_leaf_module(m):
```
How do we handle those cases when there is a control flow in the Module and we'd still like to delegate the execution to backend (given it supports that)?

jamesr66a replied:
Those should be encapsulated in functional blocks, e.g. an `LSTM` root module. From the perspective of this IR, this will look like a purely functional block, but internally it may contain control flow, which can be lowered to an accelerator natively if supported

rdzhabarov commented on torch/fx/node.py:
```diff
@@ -0,0 +1,112 @@
+# Nodes represent a definition of a value in our graph of operators.
+# For simplicitly, they also have methods defined on them so that they serve as proxy objects for _building_
+# more nodes.
+class Node:
```
* How do we handle multiple outputs from the node?

* Where do we store shape/type information for inputs/outputs?
Interested in this from the angle of lowering GraphModule to backends which require known shapes prior compilation/execution.

jamesr66a replied:
1) Uses of values from a multi-output op will appear via `getitem` calls, for example:
```
import torch

class Foo(torch.nn.Module):
    def forward(self, x):
        topk = torch.topk(x, 3)
        return torch.neg(topk[0]), topk[1]

from torch.fx import symbolic_trace
m = Foo()
gm = symbolic_trace(m)

from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name       target                                                   args         kwargs
-------------  ---------  -------------------------------------------------------  -----------  --------
placeholder    x          x                                                        ()           {}
call_function  topk_1     <built-in method topk of type object at 0x7f74b32d5760>  (x, 3)       {}
call_function  getitem    <built-in function getitem>                              (topk_1, 0)  {}
call_function  neg_1      <built-in method neg of type object at 0x7f74b32d5760>   (getitem,)   {}
call_function  getitem_1  <built-in function getitem>                              (topk_1, 1)  {}
```

2) This is up to the implementation of the analysis that puts shapes on the network, but the simplest shape propagation looks like:

```
class ShapeProp:

    def __init__(self, mod):
        self.root = mod.root
        self.graph = mod.graph
        self.state_dict = self.root.state_dict()
        self.modules = dict(self.root.named_modules())

    def propagate(self, args):
        args_iter = iter(args)
        env = {}

        def load_arg(a):
            return map_arg(a, lambda node: env[node.name])

        for node in self.graph.nodes:
            if node.op == 'placeholder':
                result = next(args_iter)
            elif node.op == 'get_param':
                result = self.state_dict[node.target]
            elif node.op == 'call_function':
                result = node.target(*load_arg(node.args), **load_arg(node.kwargs))
            elif node.op == 'call_method':
                self_obj, *args = load_arg(node.args)
                kwargs = load_arg(node.kwargs)
                result = getattr(self_obj, node.target)(*args, **kwargs)
            elif node.op == 'call_module':
                result = self.modules[node.target](*load_arg(node.args), **load_arg(node.kwargs))

            if isinstance(result, torch.Tensor):
                node.shape = result.shape
                node.dtype = result.dtype

            env[node.name] = result

        return load_arg(self.graph.result)
```

In this case all of the element types/shapes will be stored as `dtype` and `shape` attributes on each `Node` instance. Input shapes can be accessed via the attributes on the corresponding `placeholder` Node, then outputs can be accessed via the `.result` Node on `Graph`

rdzhabarov reviewed (dismissed):
At the IR level do we prohibit in-place ops as well as non tensor type args?

jamesr66a commented:
> At the IR level do we prohibit in-place ops as well as non tensor type args?

This is another scenario where we don't want to make representation choices that preclude different use cases. Passes can select whether they want to support in-place operations or not, but we're not going to limit symbolic tracing to not be able to represent those
We discussed and it probably makes sense to create a callback system s.t. passes can specify exactly what they support during tracing. We'll error out with a good stack trace if that callback fails. I can implement that in another PR

Commit:
Update on "Introduce experimental FX library"


**This feature is experimental and its stability is not currently guaranteed. Proceed at your own risk**

FX (Functional Transformations) is a toolkit for capturing and transforming functional PyTorch programs. It consists of GraphModule and a corresponding intermediate representation (IR). When GraphModule is constructed with an `nn.Module` instance as its argument, GraphModule will trace through the computation of that Module's `forward` method symbolically and record those operations in the FX intermediate representation.

```
import torch
from torch.fx import GraphModule

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.topk(torch.sum(self.linear(x + self.linear.weight).relu(), dim=-1), 3)
    
m = MyModule()
gm = GraphModule(m)
```

The Intermediate Representation centers around a 5-opcode format:

```
from tabulate import tabulate
node_specs = [[n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'name', 'target', 'args', 'kwargs']))
```

```
opcode         name           target                                                   args                kwargs
-------------  -------------  -------------------------------------------------------  ------------------  -----------
placeholder    x              x                                                        ()                  {}
get_param      linear_weight  linear.weight                                            ()                  {}
call_function  add_1          <built-in function add>                                  (x, linear_weight)  {}
call_module    linear_1       linear                                                   (add_1,)            {}
call_method    relu_2         relu                                                     [linear_1]          {}
call_function  sum_1          <built-in method sum of type object at 0x7f1c29dd36e0>   (relu_2,)           {'dim': -1}
call_function  topk_1         <built-in method topk of type object at 0x7f1c29dd36e0>  (sum_1, 3)          {}
```

The semantics are as follows:

- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` and `kwargs` are don't-care
- `get_param` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.
- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, _including the self argument_.

GraphModule automatically generates Python code for the operations it symbolically observed:

```
print(gm.src)
```

```
def forward(self, x):
    self = self.root
    linear_weight = self.linear.weight
    add_1 = x + linear_weight
    linear_1 = self.linear(add_1)
    relu_2 = linear_1.relu()
    sum_1 = torch.sum(relu_2, dim = -1)
    topk_1 = torch.topk(sum_1, 3)
    

    return topk_1
```

Because this code is valid PyTorch code, the resulting `GraphModule` can be used in any context another `nn.Module` can be used, including in TorchScript tracing/compilation.

Differential Revision: [D23006383](https://our.internmc.facebook.com/intern/diff/D23006383)

[ghstack-poisoned]

dzhulgakov commented on torch/fx/graph_module.py:
```diff
@@ -0,0 +1,80 @@
+{body}
+    return {result}
+"""
+        # print(self.src)
```
remove?

dzhulgakov commented on torch/fx/graph_module.py:
```diff
@@ -0,0 +1,80 @@
+
+# workarounds for issues in __torch_function__
+
+# WAR for __torch_function__ not handling tensor lists,
```
remove?

dzhulgakov reviewed (approved):
Awesome!
